README.md
# CalcuLab
  &emsp;&emsp;写了一个随机出算术题的程序，当做是pyqt5的基础练习。这是第一次提交项目。
  
  &emsp;&emsp;哈哈哈，不会真的有人会上GitHub还需要做这种算术题吧？
  
  ### 环境
  * python 3.8.1
  * PyQt5 5.15.0
  
  ### 使用说明
  * 运行main.py即可开始做题。
  * history.txt里是算术题历史记录，可以自己修改。
  >由于代码缺乏错误处理，很容易出现bug。~~如果你看不下去了，可以考虑训练一下你的debug能力。~~(逃)
  
  ### 部分截图
  ![截图1](https://github.com/zjjyyyk/CalcuLab/blob/main/ScreenShot/%E6%88%AA%E5%9B%BE1.png)
  ![截图2](https://github.com/zjjyyyk/CalcuLab/blob/main/ScreenShot/%E6%88%AA%E5%9B%BE2.png)





随机算术题/history.txt
Question: 682  -  158:	524
Question: 789  -  206:	583
Question: 690  -  175:	515
Question: 23  -  943:	-920
Question: 830  -  886:	-56
Question: 175  +  95:	270
Question: 924  +  618:	1542
Question: 934  -  995:	-61
Question: 837  -  478:	359
Question: 932  -  170:	762
Question: 108  +  84:	192
Question: 683  +  478:	1161
Question: 476  -  954:	*-322	*-418	*-468	-478
Question: 765  -  966:	-201
Question: 437  +  443:	880
Question: 103  +  852:	955
Question: 201  +  640:	841
Question: 125  -  107:	18
Question: 605  +  104:	709
Question: 52  +  31:	83
Question: 148  -  175:	-27
Question: 833  +  256:	1089
Question: 69  +  186:	*245	255
Question: 76  -  107:	-31
Question: 576  -  231:	345
Question: 85  -  436:	-351
Question: 514  +  601:	1115
Question: 38  +  825:	863
Question: 631  +  7:	638
Question: 612  +  329:	941
Question: 6  -  189:	-183
Question: 291  -  852:	-561
Question: 710  -  341:	369
Question: 223  -  45:	*118	178
Question: 774  +  871:	1645
Question: 31  +  77:	108
Question: 761  +  34:	795
Question: 26  -  964:	-938
Question: 638  +  315:	953
Question: 439  -  66:	*371	373
Question: 544  -  566:	-22
Question: 445  +  94:	539
Question: 787  -  660:	127
Question: 301  -  174:	*126	*125	127
Question: 899  -  873:	26
Question: 842  -  769:	73
Question: 472  +  157:	629
Question: 517  +  397:	914
Question: 355  -  903:	-548
Question: 684  -  199:	485
Question: 993  +  494:	*2117	1487
Question: 99  +  463:	562
Question: 117  -  306:	-189
Question: 712  -  548:	164
Question: 778  -  880:	-102
Question: 939  +  297:	1236
Question: 297  -  783:	-486
Question: 888  +  160:	1048
Question: 189  +  50:	239
Question: 843  +  650:	1493
Question: 650  -  393:	257
Question: 360  -  620:	-260
Question: 517  -  709:	-192
Question: 384  +  616:	1000
Question: 497  -  788:	*-311	-291
Question: 53  +  319:	372
Question: 71  -  436:	-365
Question: 185  -  452:	-267
Question: 12  -  702:	*-590	-690

Question: 55  +  865:	920
(*)	Question: 223 - 45:	*177	178
(*)	Question: 69 + 186:	*235	*245	255
(*)	Question: 301 - 174:	*227	*125	*128	127
(*)	Question: 476 - 954:	-478
(*)	Question: 439 - 66:	373
Question: 9  +  94:	103
Question: 67  -  16:	51
Question: 88  +  18:	106
Question: 93  -  11:	82
Question: 1  -  7:	-6
Question: 20  +  25:	45
Question: 100  +  40:	140
Question: 15  -  14:	1
Question: 95  -  46:	49
Question: 45  -  43:	2
Question: 29  +  21:	50
Question: 11  -  49:	-38
Question: 54  -  17:	37
Question: 82  -  37:	45
Question: 47  -  24:	23
Question: 20  -  47:	-27
Question: 33  -  21:	12
Question: 96  -  38:	58
Question: 10  -  95:	-85
Question: 0  -  72:	-72
Question: 21  -  16:	5
Question: 15  -  27:	-12
Question: 14  -  29:	*25	*-25	-15
Question: 100  -  94:	6
Question: 28  -  81:	-53
Question: 50  -  42:	8
Question: 6  -  36:	-30
Question: 36  -  22:	14
Question: 33  +  0:	33
Question: 24  +  11:	35
Question: 83  +  29:	112
Question: 45  -  60:	-15
Question: 82  -  36:	46
Question: 67  +  72:	139
Question: 17  -  80:	*-73	-63
Question: 33  -  19:	*13	*16	14
Question: 89  +  22:	111
Question: 10  +  65:	75
Question: 13  +  53:	66
Question: 0  -  4:	-4
Question: 56  -  36:	20
Question: 50  +  26:	76
Question: 30  -  59:	*-11	-29
Question: 0  -  36:	-36
Question: 56  +  42:	98
Question: 65  +  16:	81
Question: 81  +  64:	145
Question: 19  +  91:	110
Question: 37  +  68:	105
Question: 47  -  89:	-42
Question: 6  -  40:	-34
Question: 70  -  83:	-13
Question: 17  -  45:	-28
Question: 58  +  9:	67
Question: 85  +  20:	105
Question: 60  -  2:	58
Question: 38  -  29:	9
Question: 57  +  57:	114
Question: 23  -  3:	20
Question: 31  -  27:	4
Question: 94  +  79:	173
Question: 30  +  44:	74
Question: 47  -  69:	-22
Question: 8  -  86:	*-79	-78
Question: 77  -  10:	67
Question: 59  -  0:	59
Question: 48  +  66:	114
Question: 61  +  82:	143
Question: 48  -  49:	-1
Question: 68  +  100:	168
Question: 12  -  66:	-54
Question: 40  -  84:	-44
Question: 1  +  2:	3
Question: 61  +  74:	135
Question: 21  -  19:	2
Question: 35  -  29:	6
Question: 92  +  76:	168
Question: 29  +  21:	50
Question: 27  -  95:	-68
Question: 46  -  67:	-21
Question: 100  +  31:	131
Question: 5  -  29:	-24
Question: 30  +  0:	30
Question: 0  +  33:	33
Question: 90  -  88:	2
Question: 17  +  77:	94
Question: 78  +  40:	118
Question: 4  -  88:	*84	-84
Question: 40  +  82:	122
Question: 9  +  55:	64
Question: 93  +  56:	149
Question: 46  +  64:	110
Question: 44  +  85:	129
Question: 0  -  97:	-97
Question: 26  -  64:	-38
Question: 18  +  28:	46
Question: 41  +  43:	84
Question: 41  +  81:	122
Question: 82  +  33:	115
Question: 96  -  42:	54
Question: 21  +  50:	71
Question: 61  -  91:	*-20	-30
Question: 73  +  18:	91
Question: 91  +  30:	*111	121
Question: 81  +  64:	145
Question: 79  -  19:	*80	60
Question: 30  +  34:	64
Question: 9  +  71:	80
Question: 43  -  68:	-25
Question: 72  +  78:	150
Question: 10  -  25:	-15
Question: 37  +  91:	128
Question: 50  +  90:	140


随机算术题/main.py
from PyQt5.QtWidgets import QApplication, QMainWindow, QMessageBox
from PyQt5.QtCore import pyqtSlot, QUrl, QTimer
from PyQt5.QtGui import QIcon
from PyQt5.QtMultimedia import QMediaPlayer, QMediaContent, QMediaPlaylist
from UI.mainWindow import Ui_MainWindow
from setting import Setting
import sys
import random
import re
import os


class MyUi(QMainWindow):
    def __init__(self, parent=None):
        # 主窗口界面
        super().__init__(parent)
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)
        self.setFixedSize(800, 600)

        self.difficulty = 3  # 难度
        self.setWindowTitle('小学算术题 难度：%d' % self.difficulty)
        self.setWindowIcon(QIcon('res/image/windowIcon.jpg'))
        self.ui.pauseBtn.setIcon(QIcon('res/image/pause.png'))

        # 设置窗口界面
        self.setting = Setting()
        self.setting._close_signal.connect(self.on_setting_closed)

        # 音效加载
        self.playlist = QMediaPlaylist()
        self.playlist.addMedia(QMediaContent(QUrl('res/audio/correct.mp3')))
        self.playlist.addMedia(QMediaContent(QUrl('res/audio/wrong.mp3')))
        self.playlist.addMedia(QMediaContent(QUrl('res/audio/complete.mp3')))
        self.playlist.addMedia(QMediaContent(QUrl('res/audio/timeOver.mp3')))
        self.playlist.setPlaybackMode(QMediaPlaylist.CurrentItemOnce)
        self.player = QMediaPlayer()
        self.player.setPlaylist(self.playlist)
        self.player.play()

        # 定义变量
        self.answer = 0
        self.total = 0
        self.correct = 0
        self.accuracy = 0.
        self.overtime = 0
        self.tmpOutput = []

        # 保存路径
        self.fpath = 'history.txt'

        # 定时器初始化
        self.timeLimit = 30
        self.timer = QTimer(self)
        self.timer.timeout.connect(self.on_timer_timeout)
        self.currentTime = self.timeLimit
        self.ui.timeLabel.setText(str(self.currentTime))
        self.pause = False

        # 错题模式初始化
        self.mistakeBatchSize = 5
        self.mistakeMode = False
        self.exitingMistakeMode = False
        self.ui.tipLabel.clear()
        self.ui.remainingLabel.clear()
        self.questionQueue = []

        # 第一次开始运行
        self.refreshQuestion()
        self.refreshBoard()

        # 计时器开始计时
        self.timer.start(1000)

    # 刷新题目
    def refreshQuestion(self):
        # 清空self.tmpOutput以防刷新问题后将上一题也写入文件
        self.tmpOutput.clear()

        # 非错题模式随机出题
        if self.questionQueue == []:
            assert self.mistakeMode == False
            a = random.randint(0, 10 ** self.difficulty)
            b = random.randint(0, 10 ** self.difficulty)
            symbol, self.answer = (' + ', a + b) if random.random() > 0.5 else (' - ', a - b)

        # 错题模式从self.questionQueue中出题
        else:
            assert self.mistakeMode == True
            self.ui.remainingLabel.setText('剩余：{}/{}'.format(len(self.questionQueue), self.mistakeBatchSize))
            a, symbol, b = self.questionQueue.pop(0)
            self.answer = eval(a + symbol + b)
            self.tmpOutput.append('(*)')
            # 判断是否是最后一道错题
            if self.questionQueue == []:
                self.exitingMistakeMode = True

        # 将题目放入self.tmpOutput中等待写入文件
        self.tmpOutput.append('Question: ' + '{} {} {}:'.format(a, symbol, b))

        self.ui.questionLabel.setText(str(a) + symbol + str(b))
        self.ui.judgeLabel.setText('')
        self.ui.lineEdit.setFocus()

    # 更新记录板显示
    def refreshBoard(self):
        self.ui.correctLabel.setText('答对：%d/%d' % (self.correct, self.total))
        self.ui.accuracyLabel.setText('准确率：{:.0%}'.format(self.accuracy))
        self.ui.overtimeLabel.setText('超时次数：{}'.format(self.overtime))

    # 更新计时器时间
    def refreshTime(self):
        self.ui.timeLabel.setText(str(self.timeLimit))
        self.currentTime = self.timeLimit

    # 将self.tmpOutput中的字符串写入文件
    def writeFile(self):
        with open(self.fpath, 'a') as f:
            f.write('\t'.join(self.tmpOutput) + '\n')
        self.tmpOutput.clear()

    @pyqtSlot()
    def on_answerBtn_clicked(self):
        self.ui.lineEdit.setText(str(self.answer))

    @pyqtSlot()
    def on_refreshBtn_clicked(self):
        self.refreshQuestion()
        self.refreshTime()

    @pyqtSlot()
    def on_submitBtn_clicked(self):
        userInput = self.ui.lineEdit.text().strip()
        self.ui.lineEdit.clear()

        # 判断输入是否合法
        if re.match('^-?\d+$', userInput):
            # 若回答正确
            if eval(userInput) == self.answer:
                # 播放0号音效
                self.player.stop()
                self.playlist.setCurrentIndex(0)
                self.player.play()
                # UI响应
                self.ui.judgeLabel.setText('回答正确！')
                # 记录板更新变量，写入文件
                self.correct += 1
                self.total += 1
                self.accuracy = round(self.correct / self.total, 2)
                self.tmpOutput.append(userInput)
                self.writeFile()

                # 如果这是最后一个错题的话，有些地方要特殊处理
                if self.exitingMistakeMode == True:
                    # 修改状态
                    self.exitingMistakeMode = False
                    self.mistakeMode = False
                    # 播放2号音效
                    self.playlist.setCurrentIndex(2)
                    self.player.play()
                    # UI响应
                    self.ui.tipLabel.setText('完成错题模式！')
                    self.ui.remainingLabel.clear()

                # 更新问题，记录板显示，计时器时间
                self.refreshBoard()
                self.refreshQuestion()
                self.refreshTime()

            # 若回答错误
            else:
                # # 播放1号音效
                # self.player.stop()
                # self.playlist.setCurrentIndex(1)
                # self.player.play()

                # UI响应
                self.ui.judgeLabel.setText('回答错误！再思考一下！')
                # 记录板更新变量，不写入文件
                self.total += 1
                self.accuracy = round(self.correct / self.total, 2)
                self.tmpOutput.append('*' + userInput)

                # 更新记录板显示，计时器时间，不刷新问题
                self.refreshBoard()
                self.refreshTime()

    @pyqtSlot()
    def on_timer_timeout(self):
        self.currentTime -= 1
        # 若还有剩余时间
        if self.currentTime > 0:
            self.ui.timeLabel.setText(str(self.currentTime))
        # 若超时
        else:
            assert self.currentTime == 0
            # 播放3号音效
            self.player.stop()
            self.playlist.setCurrentIndex(3)
            self.player.play()

            # UI响应
            QMessageBox.information(self, '1', '超时了！')
            self.overtime += 1
            self.refreshTime()
            self.refreshBoard()

    @pyqtSlot()
    def on_pauseBtn_clicked(self):
        # 暂停
        if not self.pause:
            self.timer.stop()
            self.ui.pauseBtn.setIcon(QIcon('res/image/start.jpg'))
            self.pause = True
        # 取消暂停
        else:
            self.timer.start(1000)
            self.ui.pauseBtn.setIcon(QIcon('res/image/pause.png'))
            self.pause = False

    @pyqtSlot()
    def on_historyBtn_clicked(self):
        # 暂停
        if not self.pause:
            self.timer.stop()
            self.ui.pauseBtn.setIcon(QIcon('res/image/start.jpg'))
            self.pause = True

        # TODO 也许以后可以用QTableWidget美化输出
        # with open(self.fpath,'r') as f:
        #     content = f.read()

        # 用打开记事本的方法看历史记录
        os.system(self.fpath)

    @pyqtSlot()
    def on_mistakeModeBtn_clicked(self):
        # 开启错题模式
        self.mistakeMode = True
        # 更新计时器时间
        self.currentTime = self.timeLimit
        self.ui.timeLabel.setText(str(self.currentTime))

        # 读取历史记录中的错题（读取数不超过self.mistakeBatchSize个），并将其赋值给self.mistakeQueue
        with open(self.fpath, 'r') as f:
            lines = f.readlines()
        mistakeList = []
        for Question in lines:
            if '*' in Question:
                questionText = Question.split(':')[1].strip()
                assert len(questionText.split()) == 3
                mistakeList.append(questionText.split())
                if len(mistakeList) == self.mistakeBatchSize:
                    break
        assert len(mistakeList) <= self.mistakeBatchSize
        random.shuffle(mistakeList)  # 随机打乱这些错题
        self.questionQueue = mistakeList

        # UI响应
        self.ui.tipLabel.setText('正在重做错题模式')
        self.ui.lineEdit.clear()

        # 刷新问题
        self.refreshQuestion()

    @pyqtSlot()
    def on_settingBtn_clicked(self):
        # 暂停
        if not self.pause:
            self.timer.stop()
            self.ui.pauseBtn.setIcon(QIcon('res/image/start.jpg'))
            self.pause = True

        self.setting.show()

    @pyqtSlot()
    def on_setting_closed(self):
        # 将设置值赋予主窗口
        self.difficulty, self.timeLimit, self.mistakeBatchSize = \
            self.setting.difficultyValue, self.setting.timeLimitValue, self.setting.mistakeBatchSizeValue


if __name__ == '__main__':
    app = QApplication(sys.argv)
    myApp = MyUi()
    myApp.show()
    sys.exit(app.exec_())


随机算术题/setting.py
from PyQt5.QtWidgets import QWidget
from PyQt5.QtCore import pyqtSlot, pyqtSignal
from PyQt5.QtGui import QIcon
from UI.settingWidget import Ui_Form


class Setting(QWidget):
    # 定义关闭时的信号
    _close_signal = pyqtSignal()

    def __init__(self, parent=None):
        super().__init__(parent=parent)
        self.ui = Ui_Form()
        self.ui.setupUi(self)
        self.setFixedSize(500, 380)
        self.setWindowTitle('设置')
        self.setWindowIcon(QIcon('res/image/setting.jpg'))

        # 默认值
        self.difficultyValue_default = 3
        self.timeLimitValue_default = 30
        self.mistakeBatchSizeValue_default = 5

        self.difficultyValue = self.difficultyValue_default
        self.timeLimitValue = self.timeLimitValue_default
        self.mistakeBatchSizeValue = self.mistakeBatchSizeValue_default

    @pyqtSlot(int)
    def on_difficultySlider_valueChanged(self, value):
        self.difficultyValue = value
        self.ui.difficultyLabel.setText('难度：{}'.format(self.difficultyValue))

    @pyqtSlot(int)
    def on_timeLimitSlider_valueChanged(self, value):
        self.timeLimitValue = value
        self.ui.timeLimitLabel.setText('时间：{}'.format(self.timeLimitValue))

    @pyqtSlot(int)
    def on_mistakeBatchSizeSlider_valueChanged(self, value):
        self.mistakeBatchSizeValue = value
        self.ui.mistakeBatchSizeLabel.setText('错题数：{}'.format(self.mistakeBatchSizeValue))

    @pyqtSlot()
    def on_resetBtn_clicked(self):
        self.ui.difficultySlider.setValue(self.difficultyValue_default)
        self.ui.timeLimitSlider.setValue(self.timeLimitValue_default)
        self.ui.mistakeBatchSizeSlider.setValue(self.mistakeBatchSizeValue_default)

    @pyqtSlot()
    def on_yesBtn_clicked(self):
        self.close()

    # 重写关闭事件：发射关闭信号以便让主窗口知道自己关闭了
    def closeEvent(self, QCloseEvent):
        self._close_signal.emit()


随机算术题/UI/settingWidget.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'settingWidget.ui'
#
# Created by: PyQt5 UI code generator 5.15.0
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_Form(object):
    def setupUi(self, Form):
        Form.setObjectName("Form")
        Form.resize(431, 344)
        self.resetBtn = QtWidgets.QPushButton(Form)
        self.resetBtn.setGeometry(QtCore.QRect(20, 280, 111, 28))
        self.resetBtn.setObjectName("resetBtn")
        self.yesBtn = QtWidgets.QPushButton(Form)
        self.yesBtn.setGeometry(QtCore.QRect(240, 280, 93, 28))
        self.yesBtn.setObjectName("yesBtn")
        self.widget = QtWidgets.QWidget(Form)
        self.widget.setGeometry(QtCore.QRect(20, 30, 351, 221))
        self.widget.setObjectName("widget")
        self.gridLayout = QtWidgets.QGridLayout(self.widget)
        self.gridLayout.setContentsMargins(0, 0, 0, 0)
        self.gridLayout.setObjectName("gridLayout")
        self.timeLimitLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setPointSize(13)
        self.timeLimitLabel.setFont(font)
        self.timeLimitLabel.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.timeLimitLabel.setObjectName("timeLimitLabel")
        self.gridLayout.addWidget(self.timeLimitLabel, 1, 0, 1, 1)
        self.difficultyLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setPointSize(13)
        self.difficultyLabel.setFont(font)
        self.difficultyLabel.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.difficultyLabel.setObjectName("difficultyLabel")
        self.gridLayout.addWidget(self.difficultyLabel, 0, 0, 1, 1)
        self.timeLimitSlider = QtWidgets.QSlider(self.widget)
        self.timeLimitSlider.setMinimum(5)
        self.timeLimitSlider.setMaximum(100)
        self.timeLimitSlider.setProperty("value", 30)
        self.timeLimitSlider.setOrientation(QtCore.Qt.Horizontal)
        self.timeLimitSlider.setObjectName("timeLimitSlider")
        self.gridLayout.addWidget(self.timeLimitSlider, 1, 1, 1, 1)
        self.mistakeBatchSizeLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setPointSize(13)
        self.mistakeBatchSizeLabel.setFont(font)
        self.mistakeBatchSizeLabel.setAlignment(QtCore.Qt.AlignRight|QtCore.Qt.AlignTrailing|QtCore.Qt.AlignVCenter)
        self.mistakeBatchSizeLabel.setObjectName("mistakeBatchSizeLabel")
        self.gridLayout.addWidget(self.mistakeBatchSizeLabel, 2, 0, 1, 1)
        self.mistakeBatchSizeSlider = QtWidgets.QSlider(self.widget)
        self.mistakeBatchSizeSlider.setMinimum(1)
        self.mistakeBatchSizeSlider.setMaximum(20)
        self.mistakeBatchSizeSlider.setProperty("value", 5)
        self.mistakeBatchSizeSlider.setOrientation(QtCore.Qt.Horizontal)
        self.mistakeBatchSizeSlider.setObjectName("mistakeBatchSizeSlider")
        self.gridLayout.addWidget(self.mistakeBatchSizeSlider, 2, 1, 1, 1)
        self.difficultySlider = QtWidgets.QSlider(self.widget)
        self.difficultySlider.setMinimum(1)
        self.difficultySlider.setMaximum(8)
        self.difficultySlider.setProperty("value", 3)
        self.difficultySlider.setOrientation(QtCore.Qt.Horizontal)
        self.difficultySlider.setObjectName("difficultySlider")
        self.gridLayout.addWidget(self.difficultySlider, 0, 1, 1, 1)

        self.retranslateUi(Form)
        QtCore.QMetaObject.connectSlotsByName(Form)

    def retranslateUi(self, Form):
        _translate = QtCore.QCoreApplication.translate
        Form.setWindowTitle(_translate("Form", "Form"))
        self.resetBtn.setText(_translate("Form", "恢复默认设置"))
        self.yesBtn.setText(_translate("Form", "确定"))
        self.yesBtn.setShortcut(_translate("Form", "Ctrl+S"))
        self.timeLimitLabel.setText(_translate("Form", "时间：30"))
        self.difficultyLabel.setText(_translate("Form", "难度：3"))
        self.mistakeBatchSizeLabel.setText(_translate("Form", "错题数：5"))


随机算术题/UI/mainWindow.py
# -*- coding: utf-8 -*-

# Form implementation generated from reading ui file 'mainWindow.ui'
#
# Created by: PyQt5 UI code generator 5.15.0
#
# WARNING: Any manual changes made to this file will be lost when pyuic5 is
# run again.  Do not edit this file unless you know what you are doing.


from PyQt5 import QtCore, QtGui, QtWidgets


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.resize(800, 600)
        sizePolicy = QtWidgets.QSizePolicy(QtWidgets.QSizePolicy.Fixed, QtWidgets.QSizePolicy.Fixed)
        sizePolicy.setHorizontalStretch(0)
        sizePolicy.setVerticalStretch(0)
        sizePolicy.setHeightForWidth(MainWindow.sizePolicy().hasHeightForWidth())
        MainWindow.setSizePolicy(sizePolicy)
        font = QtGui.QFont()
        font.setStrikeOut(False)
        MainWindow.setFont(font)
        MainWindow.setStyleSheet("")
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.questionLabel = QtWidgets.QLabel(self.centralwidget)
        self.questionLabel.setGeometry(QtCore.QRect(150, 40, 471, 161))
        font = QtGui.QFont()
        font.setFamily("Arial")
        font.setPointSize(30)
        font.setStrikeOut(False)
        self.questionLabel.setFont(font)
        self.questionLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.questionLabel.setObjectName("questionLabel")
        self.widget = QtWidgets.QWidget(self.centralwidget)
        self.widget.setGeometry(QtCore.QRect(540, 350, 291, 191))
        self.widget.setObjectName("widget")
        self.verticalLayout = QtWidgets.QVBoxLayout(self.widget)
        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
        self.verticalLayout.setObjectName("verticalLayout")
        self.correctLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setFamily("微软雅黑")
        font.setPointSize(15)
        font.setStrikeOut(False)
        self.correctLabel.setFont(font)
        self.correctLabel.setStyleSheet("color:blue")
        self.correctLabel.setObjectName("correctLabel")
        self.verticalLayout.addWidget(self.correctLabel)
        self.accuracyLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setFamily("微软雅黑")
        font.setPointSize(15)
        font.setStrikeOut(False)
        self.accuracyLabel.setFont(font)
        self.accuracyLabel.setStyleSheet("color:red")
        self.accuracyLabel.setObjectName("accuracyLabel")
        self.verticalLayout.addWidget(self.accuracyLabel)
        self.overtimeLabel = QtWidgets.QLabel(self.widget)
        font = QtGui.QFont()
        font.setFamily("微软雅黑")
        font.setPointSize(15)
        font.setStrikeOut(False)
        self.overtimeLabel.setFont(font)
        self.overtimeLabel.setStyleSheet("color:grey")
        self.overtimeLabel.setObjectName("overtimeLabel")
        self.verticalLayout.addWidget(self.overtimeLabel)
        self.widget_2 = QtWidgets.QWidget(self.centralwidget)
        self.widget_2.setGeometry(QtCore.QRect(210, 170, 343, 151))
        self.widget_2.setObjectName("widget_2")
        self.gridLayout = QtWidgets.QGridLayout(self.widget_2)
        self.gridLayout.setContentsMargins(0, 0, 0, 0)
        self.gridLayout.setObjectName("gridLayout")
        self.lineEdit = QtWidgets.QLineEdit(self.widget_2)
        font = QtGui.QFont()
        font.setPointSize(20)
        font.setStrikeOut(False)
        self.lineEdit.setFont(font)
        self.lineEdit.setText("")
        self.lineEdit.setObjectName("lineEdit")
        self.gridLayout.addWidget(self.lineEdit, 2, 2, 1, 1)
        self.submitBtn = QtWidgets.QPushButton(self.widget_2)
        font = QtGui.QFont()
        font.setPointSize(9)
        font.setStrikeOut(False)
        self.submitBtn.setFont(font)
        self.submitBtn.setObjectName("submitBtn")
        self.gridLayout.addWidget(self.submitBtn, 3, 2, 1, 1)
        self.judgeLabel = QtWidgets.QLabel(self.widget_2)
        font = QtGui.QFont()
        font.setPointSize(10)
        font.setStrikeOut(False)
        self.judgeLabel.setFont(font)
        self.judgeLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.judgeLabel.setObjectName("judgeLabel")
        self.gridLayout.addWidget(self.judgeLabel, 1, 2, 1, 1)
        self.label = QtWidgets.QLabel(self.widget_2)
        font = QtGui.QFont()
        font.setPointSize(20)
        font.setBold(True)
        font.setWeight(75)
        font.setStrikeOut(False)
        self.label.setFont(font)
        self.label.setObjectName("label")
        self.gridLayout.addWidget(self.label, 2, 0, 1, 1)
        self.answerBtn = QtWidgets.QPushButton(self.centralwidget)
        self.answerBtn.setGeometry(QtCore.QRect(600, 100, 111, 41))
        self.answerBtn.setObjectName("answerBtn")
        self.widget_3 = QtWidgets.QWidget(self.centralwidget)
        self.widget_3.setGeometry(QtCore.QRect(30, 30, 141, 91))
        self.widget_3.setObjectName("widget_3")
        self.horizontalLayout = QtWidgets.QHBoxLayout(self.widget_3)
        self.horizontalLayout.setContentsMargins(0, 0, 0, 0)
        self.horizontalLayout.setObjectName("horizontalLayout")
        self.timeLabel = QtWidgets.QLabel(self.widget_3)
        font = QtGui.QFont()
        font.setPointSize(30)
        font.setStrikeOut(False)
        self.timeLabel.setFont(font)
        self.timeLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.timeLabel.setObjectName("timeLabel")
        self.horizontalLayout.addWidget(self.timeLabel)
        self.pauseBtn = QtWidgets.QToolButton(self.widget_3)
        self.pauseBtn.setStyleSheet("background-color:transparent")
        self.pauseBtn.setIconSize(QtCore.QSize(50, 50))
        self.pauseBtn.setObjectName("pauseBtn")
        self.horizontalLayout.addWidget(self.pauseBtn)
        self.widget_4 = QtWidgets.QWidget(self.centralwidget)
        self.widget_4.setGeometry(QtCore.QRect(50, 410, 421, 121))
        self.widget_4.setObjectName("widget_4")
        self.gridLayout_2 = QtWidgets.QGridLayout(self.widget_4)
        self.gridLayout_2.setContentsMargins(0, 0, 0, 0)
        self.gridLayout_2.setObjectName("gridLayout_2")
        self.tipLabel = QtWidgets.QLabel(self.widget_4)
        font = QtGui.QFont()
        font.setFamily("华光细圆_CNKI")
        font.setPointSize(10)
        font.setBold(True)
        font.setWeight(75)
        font.setStrikeOut(False)
        self.tipLabel.setFont(font)
        self.tipLabel.setStyleSheet("color:orange;")
        self.tipLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.tipLabel.setObjectName("tipLabel")
        self.gridLayout_2.addWidget(self.tipLabel, 0, 3, 1, 1)
        self.remainingLabel = QtWidgets.QLabel(self.widget_4)
        font = QtGui.QFont()
        font.setFamily("华光细圆_CNKI")
        font.setPointSize(10)
        font.setBold(True)
        font.setWeight(75)
        font.setStrikeOut(False)
        self.remainingLabel.setFont(font)
        self.remainingLabel.setStyleSheet("color:orange;")
        self.remainingLabel.setAlignment(QtCore.Qt.AlignCenter)
        self.remainingLabel.setObjectName("remainingLabel")
        self.gridLayout_2.addWidget(self.remainingLabel, 1, 3, 1, 1)
        self.mistakeModeBtn = QtWidgets.QPushButton(self.widget_4)
        self.mistakeModeBtn.setObjectName("mistakeModeBtn")
        self.gridLayout_2.addWidget(self.mistakeModeBtn, 2, 3, 1, 1)
        self.historyBtn = QtWidgets.QPushButton(self.widget_4)
        self.historyBtn.setObjectName("historyBtn")
        self.gridLayout_2.addWidget(self.historyBtn, 2, 1, 1, 1)
        self.settingBtn = QtWidgets.QPushButton(self.widget_4)
        self.settingBtn.setObjectName("settingBtn")
        self.gridLayout_2.addWidget(self.settingBtn, 2, 0, 1, 1)
        self.refreshBtn = QtWidgets.QPushButton(self.centralwidget)
        self.refreshBtn.setGeometry(QtCore.QRect(600, 160, 111, 41))
        self.refreshBtn.setObjectName("refreshBtn")
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 800, 26))
        self.menubar.setObjectName("menubar")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "MainWindow"))
        self.questionLabel.setText(_translate("MainWindow", "1000 - 1000"))
        self.correctLabel.setText(_translate("MainWindow", "答对：10/10"))
        self.accuracyLabel.setText(_translate("MainWindow", "准确率：100%"))
        self.overtimeLabel.setText(_translate("MainWindow", "超时次数：0次"))
        self.submitBtn.setText(_translate("MainWindow", "确定"))
        self.submitBtn.setShortcut(_translate("MainWindow", "Return"))
        self.judgeLabel.setText(_translate("MainWindow", "回答正确！"))
        self.label.setText(_translate("MainWindow", "答案是："))
        self.answerBtn.setText(_translate("MainWindow", "查看正确答案"))
        self.timeLabel.setText(_translate("MainWindow", "30"))
        self.pauseBtn.setText(_translate("MainWindow", "暂停"))
        self.tipLabel.setText(_translate("MainWindow", "正在重做错题模式"))
        self.remainingLabel.setText(_translate("MainWindow", "剩余错题：5/5"))
        self.mistakeModeBtn.setText(_translate("MainWindow", "重做错题"))
        self.historyBtn.setText(_translate("MainWindow", "查看所有历史记录"))
        self.settingBtn.setText(_translate("MainWindow", "设置"))
        self.refreshBtn.setText(_translate("MainWindow", "刷新题目"))
